File: user-service/src/main/java/rs/edu/raf/service/impl/CompanyServiceImpl.java
package rs.edu.raf.service.impl;

import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;
import rs.edu.raf.dto.CompanyCreateDTO;
import rs.edu.raf.dto.CompanyDTO;
import rs.edu.raf.exceptions.CompanyNotFoundException;
import rs.edu.raf.mapper.CompanyMapper;
import rs.edu.raf.model.user.Company;
import rs.edu.raf.repository.CompanyRepository;
import rs.edu.raf.service.CompanyService;

import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
public class CompanyServiceImpl implements CompanyService {
    private CompanyRepository companyRepository;
    private CompanyMapper companyMapper;

    @Override
    public CompanyDTO createCompany(CompanyCreateDTO companyCreateDTO) {
        return companyMapper.companyToCompanyDTO(companyRepository.save(companyMapper.companyCreateDTOtoCompany(companyCreateDTO)));
    }

    @Override
    public CompanyDTO findCompanyById(Long id) {
        return companyMapper.companyToCompanyDTO(companyRepository.findById(id)
                .orElseThrow(()->new CompanyNotFoundException("Company with ID " + id + " not found!")));
    }

    @Override
    public List<CompanyDTO> getAllCompanies() {
        return companyRepository.findAll().stream().map(companyMapper::companyToCompanyDTO).toList();
    }

    @Override
    public Long addBankAccountToCompany(Integer TIN, Long bankAccountNumber) {
        Optional<Company> optionalCompany = companyRepository.findByTIN(TIN);

        if(optionalCompany.isEmpty()) {
            return null;
        }

        Company company = optionalCompany.get();

        if(company.getAccountNumbers() == null){
            company.setAccountNumbers(bankAccountNumber.toString());
            companyRepository.save(company);
            return company.getId();
        }

        if(company.getAccountNumbers().contains(bankAccountNumber.toString())){
            return null;
        }

        company.setAccountNumbers(company.getAccountNumbers() + "," + bankAccountNumber);
        companyRepository.save(company);
        return company.getId();
    }
}

File: user-service-kotlin/src/main/kotlin/org/example/userservicekotlin/service/impl/CompanyServiceImpl.kt
package org.example.userservicekotlin.service.impl

import org.example.userservicekotlin.dto.CompanyCreateDTO
import org.example.userservicekotlin.dto.CompanyDTO
import org.example.userservicekotlin.mapper.CompanyMapper
import org.example.userservicekotlin.model.user.Company
import org.example.userservicekotlin.repository.CompanyRepository
import org.example.userservicekotlin.service.CompanyService
import org.example.userservicekotlin.exception.CompanyNotFoundException
import org.springframework.stereotype.Service
import java.util.*

@Service
class CompanyServiceImpl(val companyRepository: CompanyRepository, val companyMapper: CompanyMapper) : CompanyService {
    override fun createCompany(companyCreateDTO: CompanyCreateDTO): CompanyDTO {
        return companyMapper.companyToCompanyDTO(
            companyRepository.save(
                companyMapper.companyCreateDTOtoCompany(
                    companyCreateDTO
                )
            )
        )
    }

    override fun findCompanyById(id: Long): CompanyDTO {
        return companyMapper.companyToCompanyDTO(
            companyRepository.findById(id)
                .orElseThrow { CompanyNotFoundException("Company with ID $id not found!") })
    }

    override fun getAllCompanies(): List<CompanyDTO> {
        return companyRepository.findAll().stream().map(companyMapper::companyToCompanyDTO).toList()
    }

    override fun addBankAccountToCompany(TIN: Int, bankAccountNumber: Long): Long? {
        val optionalCompany: Optional<Company> = companyRepository.findByTIN(TIN)

        if (optionalCompany.isEmpty()) {
            return null
        }

        val company: Company = optionalCompany.get()

        if (company.accountNumbers == null) {
            company.accountNumbers = bankAccountNumber.toString()
            companyRepository.save(company)
            return company.id
        }

        if (company.accountNumbers!!.contains(bankAccountNumber.toString())) {
            return null
        }

        company.accountNumbers = company.accountNumbers + "," + bankAccountNumber
        companyRepository.save(company)
        return company.id
    }
}